oak_runtime/src/node/grpc/invocation.rs

```diff
@@ -16,70 +16,54 @@
 
 //! Functionality for handling gRPC method invocations.
 
+pub use crate::proto::oak::invocation::GrpcInvocation as Invocation;
 use crate::{
-    io::{Receiver, ReceiverExt, Sender, SenderExt},
-    NodeMessage, RuntimeProxy,
+    io::{ReceiverExt, SenderExt},
+    RuntimeProxy,
 };
 use log::error;
 use oak_abi::OakStatus;
-use oak_io::{
-    handle::{ReadHandle, WriteHandle},
-    Decodable, OakError,
-};
+use oak_io::OakError;
 use oak_services::proto::oak::encap::{GrpcRequest, GrpcResponse};
 
-/// A gRPC invocation, consisting of exactly two channels: one to read incoming requests from the
-/// client (wrapped in a [`Receiver`]), and one to write outgoing responses to the client (wrapped
-/// in a [`Sender`]).
-pub struct Invocation {
-    receiver: Receiver<GrpcRequest>,
-    sender: Sender<GrpcResponse>,
-}
-
-impl Decodable for Invocation {
-    fn decode(message: &NodeMessage) -> Result<Self, OakError> {
-        if !message.bytes.is_empty() {
-            error!("Non-empty data field");
-            return Err(OakStatus::ErrInternal.into());
-        }
-        if message.handles.len() != 2 {
-            error!(
-                "Incorrect number of handles received: {} (expected: 2)",
-                message.handles.len()
-            );
-            return Err(OakStatus::ErrInternal.into());
-        }
-        Ok(Self {
-            receiver: Receiver::new(ReadHandle {
-                handle: message.handles[0],
-            }),
-            sender: Sender::new(WriteHandle {
-                handle: message.handles[1],
-            }),
-        })
-    }
-}
-
 impl Invocation {
```

Comment:
Could this and a similar implementation of HttpInvocation in `sdk/rust/oak/src/http/mod.rs` be unified into a generic implementation?

Reply:
I don't think this can be done easily. This `GrpcInvocation` implementation is in the Runtime and the implementation of the `HttpInvocation` is in the SDK. The Runtime and SDK use different versions of SenderExt and ReceiverExt with different signatures for `send`, `receive` and `close` to support different mechanisms for interacting with handles depending on whether the code runs in a WASM node or not.

Reply:
hmm! I did not realize that we have two different versions of `SenderExt` and `ReceiverExt `.